<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Renaming · VKCComputing</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://Klepac-Ceraj-Lab.github.io/VKCComputing.jl/examples/renaming_files/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">VKCComputing</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li class="is-active"><a class="tocitem" href>Renaming</a><ul class="internal"><li><a class="tocitem" href="#Motivation"><span>Motivation</span></a></li><li><a class="tocitem" href="#Getting-current-data"><span>Getting current data</span></a></li><li><a class="tocitem" href="#Building-the-renaming-map"><span>Building the renaming map</span></a></li><li><a class="tocitem" href="#Finding-files-to-rename"><span>Finding files to rename</span></a></li><li><a class="tocitem" href="#Renaming"><span>Renaming</span></a></li><li><a class="tocitem" href="#Dealing-with-Ambiguities"><span>Dealing with Ambiguities</span></a></li><li><a class="tocitem" href="#Conclusion"><span>Conclusion</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Renaming</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Renaming</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Klepac-Ceraj-Lab/VKCComputing.jl/blob/main/docs/src/examples/renaming_files.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Renaming-files-(from-biospecimen-IDs-and-aliases)"><a class="docs-heading-anchor" href="#Renaming-files-(from-biospecimen-IDs-and-aliases)">Renaming files (from biospecimen IDs and aliases)</a><a id="Renaming-files-(from-biospecimen-IDs-and-aliases)-1"></a><a class="docs-heading-anchor-permalink" href="#Renaming-files-(from-biospecimen-IDs-and-aliases)" title="Permalink"></a></h1><h2 id="Motivation"><a class="docs-heading-anchor" href="#Motivation">Motivation</a><a id="Motivation-1"></a><a class="docs-heading-anchor-permalink" href="#Motivation" title="Permalink"></a></h2><p>After changing the way that we label samples, we sometimes need to update a previous file-name or table column name to reflect the new system. This code will not produce the same outputs anymore, since the state of drives and the database has changed in the meantime. It&#39;s meant more as documentation of the process.</p><h2 id="Getting-current-data"><a class="docs-heading-anchor" href="#Getting-current-data">Getting current data</a><a id="Getting-current-data-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-current-data" title="Permalink"></a></h2><p>The first thing to do in most projects is to load the airtable database into memory. If you want to guarantee that you have the most recent version of any particular table, use the <code>update</code> argument of <a href="../../#VKCComputing.LocalBase"><code>LocalBase</code></a>.</p><pre><code class="language-julia-repl hljs">julia&gt; using VKCComputing, Dates

julia&gt; base = LocalBase(; update=[&quot;Biospecimens&quot;=&gt; Week(1), &quot;Projects&quot;=&gt; true]);
[ Info: Table Aliases does not need updating updating. Use `update = true` or `update = {shorter interval}` to override
[ Info: Loading records from local JSON file
[ Info: Table Biospecimens does not need updating updating. Use `update = true` or `update = {shorter interval}` to override
[ Info: Loading records from local JSON file
#...</code></pre><p>Let&#39;s say I want to find any samples from the ECHO project that need to be updated. I don&#39;t remember how the ECHO project is referred to in the database, so I need to check what the <code>uid</code>s from the &quot;Projects&quot; table are:</p><pre><code class="language-julia-repl hljs">julia&gt; uids(base, &quot;Projects&quot;) # or `uids(base[&quot;Projects&quot;])`
2-element Dictionaries.Indices{String}
 &quot;resonance&quot;
 &quot;khula&quot;</code></pre><p>Looks like it&#39;s &quot;resonance&quot;! Next, I&#39;ll get all of the samples associated with that project. There are a couple of ways to do that; I&#39;ll use a somewhat roundabout way to show off a couple of features:</p><pre><code class="language-julia-repl hljs">julia&gt; proj = base[&quot;Projects&quot;, &quot;resonance&quot;]; # get the project record

julia&gt; keys(proj); # look at what fields are available... I want :Subjects
(:uid, :name, :Visits, :Subjects)

julia&gt; first(proj[:Subjects], 5) # verify that these are record hashes
5-element Vector{String}:
 &quot;recF3MoP0RZ4EqRrh&quot;
 &quot;recnZclKoXvxPbARR&quot;
 &quot;recW3HYwsZPYt8rtD&quot;
 &quot;recmljdPsUa55j2BS&quot;
 &quot;recaRvIXKLtDNQBOb&quot;</code></pre><p>Once I have a list of records hashes, I can use them to pull records directly:</p><pre><code class="language-julia-repl hljs">julia&gt; subjects = base[proj[:Subjects]]
770-element Vector{Airtable.AirRecord}:
 Airtable.AirRecord(&quot;recF3MoP0RZ4EqRrh&quot;, AirTable(&quot;Subjects&quot;), (uid = &quot;1255&quot;, Biospecimens = [&quot;recnfhuOaRwSfOLOq&quot;, &quot;recrSIHTxVxj3JXav&quot;, &quot;rec7j3VFnZb0Uue5k&quot;, &quot;recrfP0SIqxTjk0Vk&quot;, &quot;recLsUbcOI32ZjrSN&quot;, &quot;rec1Ai2Nz0yCmpLpa&quot;, &quot;receGDNvbRuQpiCNQ&quot;, &quot;recPVakgZpe01B9ZJ&quot;, &quot;rec5iE5o92ManNgDG&quot;], project = [&quot;recOlnbWbRoJdy4Nq&quot;]))
 Airtable.AirRecord(&quot;recnZclKoXvxPbARR&quot;, AirTable(&quot;Subjects&quot;), (uid = &quot;0672&quot;, Biospecimens = [&quot;recnKOG7QwOKCG8Tk&quot;, &quot;recisA4suwsc4iq3U&quot;, &quot;rec98YN0dSis3lxbQ&quot;, &quot;recV29l0ADxUL2xk5&quot;, &quot;recLRCssF2ZBC8suZ&quot;, &quot;rec46VGKOVth2SQ6U&quot;, &quot;recDUj6DDLyjNSbki&quot;, &quot;reclzmMkTJiWFCJWI&quot;, &quot;reczBSh8Vav6D0qQy&quot;, &quot;reclUM1eREFipQ6ND&quot;], project = [&quot;recOlnbWbRoJdy4Nq&quot;]))
 #...</code></pre><p>Now, I can get all of the biospecimens associated with these subjects:</p><pre><code class="language-julia-repl hljs">julia&gt; mapreduce(rec-&gt; get(rec, :Biospecimens, []), vcat, subjects)
3317-element Vector{String}:
 &quot;recnfhuOaRwSfOLOq&quot;
 &quot;recrSIHTxVxj3JXav&quot;
 &quot;rec7j3VFnZb0Uue5k&quot;
 &quot;recrfP0SIqxTjk0Vk&quot;
#...</code></pre><p>Then, I can use these record hashes to pull the biospecimen records:</p><pre><code class="language-julia-repl hljs">julia&gt; biosp = base[mapreduce(rec-&gt; get(rec, :Biospecimens, []), vcat, subjects)]
3317-element Vector{Airtable.AirRecord}:
 Airtable.AirRecord(&quot;recnfhuOaRwSfOLOq&quot;, AirTable(&quot;Biospecimens&quot;), (uid = &quot;FE50074&quot;, subject = [&quot;recF3MoP0RZ4EqRrh&quot;], collection_buffer = [&quot;recxsTHmTS84TBMPF&quot;], aliases = [&quot;recP0jpfk49JXtVp2&quot;], project = [&quot;recOlnbWbRoJdy4Nq&quot;]))
 Airtable.AirRecord(&quot;recrSIHTxVxj3JXav&quot;, AirTable(&quot;Biospecimens&quot;), (uid = &quot;FE01868&quot;, subject = [&quot;recF3MoP0RZ4EqRrh&quot;], collection = 3, collection_buffer = [&quot;recxsTHmTS84TBMPF&quot;], visit = [&quot;recjx3Tb7wf6WXO6t&quot;], aliases = [&quot;recsqOcdY54CU5hQT&quot;], project = [&quot;recOlnbWbRoJdy4Nq&quot;]))
#...</code></pre><p>A more straightforwards approach would have been to look at all of the biospecimen records, and filter on the ones where <code>Project</code> had the id hash for the <code>&quot;resonance&quot;</code> project.</p><h2 id="Building-the-renaming-map"><a class="docs-heading-anchor" href="#Building-the-renaming-map">Building the renaming map</a><a id="Building-the-renaming-map-1"></a><a class="docs-heading-anchor-permalink" href="#Building-the-renaming-map" title="Permalink"></a></h2><p>Now that we have all of the biospecimens for ECHO, let&#39;s just get the ones that have been shotgun sequenced, and identify the <code>SequencingPrep</code> records associated with them.</p><pre><code class="language-julia-repl hljs">julia&gt; mapreduce(keys, union, biosp)
8-element Vector{Symbol}:
 :uid
 :subject
 :collection_buffer
 :aliases
 :project
 :collection
 :visit
 :seqprep

julia&gt; filter!(rec-&gt; haskey(rec, :seqprep), biosp);</code></pre><p>Now, we want to build a mapping of biospecimen =&gt; seqprep ID. I&#39;ll store this as rows in a <code>DataFrame</code>.</p><pre><code class="language-julia-repl hljs">julia&gt; rndf = DataFrame(mapreduce(vcat, biosp) do rec
    rows = [(; seqname = base[id][:uid], oldname = rec[:uid]) for id in rec[:seqprep]]
end);</code></pre><p>It is in principle possible for a single biospecimen ID to refer to multiple seqprep IDs. Let&#39;s check that:</p><pre><code class="language-julia-repl hljs">julia&gt; transform!(groupby(rndf, :oldname), &quot;seqname&quot;=&gt; length =&gt; &quot;n_seqs&quot;);

julia&gt; subset(rndf, &quot;n_seqs&quot;=&gt; ByRow(&gt;(1)))
6×3 DataFrame
 Row │ seqname   oldname  n_seqs
     │ String    String   Int64
─────┼───────────────────────────
   1 │ SEQ02303  FG02294       2
   2 │ SEQ02303  FG02294       2
   3 │ SEQ01232  FE01105       2
   4 │ SEQ01110  FG00016       2
   5 │ SEQ01110  FG00016       2
   6 │ SEQ01232  FE01105       2</code></pre><p>Now, we&#39;ll build a map of <code>oldname</code> -&gt; <code>seqname</code>.</p><pre><code class="language-julia-repl hljs">rnmap = dictionary(zip(rndf.oldname, rndf.seqname))
1799-element Dictionary{String, String}
 &quot;FG50159&quot; │ &quot;SEQ01107&quot;
 &quot;FG50160&quot; │ &quot;SEQ01108&quot;
 &quot;FG00846&quot; │ &quot;SEQ01371&quot;
 #...</code></pre><h2 id="Finding-files-to-rename"><a class="docs-heading-anchor" href="#Finding-files-to-rename">Finding files to rename</a><a id="Finding-files-to-rename-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-files-to-rename" title="Permalink"></a></h2><p>I&#39;m running this on <code>hopper</code>, so files are contained in the <code>/grace</code> drive, as well as some other places. But we&#39;ll start there.</p><p>The first thing to do is find all of the files that could plausibly be in <code>ECHO</code>. They all fit the pattern <code>r&quot;F[EG]\d{5}</code> - that is, &quot;FE or &quot;FG&quot; followed by 5 numbers (<code>\d</code> stands for &quot;digit&quot;). Let&#39;s make sure that&#39;s true for our rename map:</p><pre><code class="nohighlight hljs">julia&gt; all(k-&gt; contains(k, r&quot;F[EG]\d{5}&quot;), keys(rnmap))
true</code></pre><p>So now we&#39;ll recurse through the directory, saving any files that fit the pattern. While we&#39;re at it, we can pull out some relevant info and push it into a <code>DataFrame</code>.</p><pre><code class="language-julia-repl hljs">julia&gt; filedf = DataFrame()

julia&gt;  for (root, dir, files) in walkdir(&quot;/grace/sequencing/processed/mgx&quot;)
            for file in files
                m = match(r&quot;^(F[EG]\d{5})_(S\d+)_&quot;, file)
                isnothing(m) &amp;&amp; continue
                newname = get(rnmap, m[1], nothing)
                if isnothing(newname)
                    @warn &quot;$(m[1]) matches the regex, but doesn&#39;t have a new name&quot;
                    continue
                end
                push!(filedf, (;
                    oldname = m[1], newname, snum = m[2], oldpath = joinpath(root, file), newpath = joinpath(root, replace(file, m[1]=&gt;newname))
                ))
            end
        end

julia&gt; first(filedf, 5)
5×5 DataFrame
 Row │ oldname    newname   snum       oldpath                            newpath
     │ SubStrin…  String    SubStrin…  String                             String
─────┼──────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ FE01063    SEQ02371  S93        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…
   2 │ FE01063    SEQ02371  S93        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…
   3 │ FE01063    SEQ02371  S93        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…
   4 │ FE01064    SEQ00905  S10        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…
   5 │ FE01064    SEQ00905  S10        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…
</code></pre><p>Now, we want to double check that all of the <code>oldname</code>s are associated with the same <code>snum</code> to avoid ambiguities. Here, we group by <code>oldname</code>, and then look at the <code>snum</code> column to ensure there is only 1 (we check that the length of unique elements in that column is 1).</p><pre><code class="nohighlight hljs">julia&gt; transform!(groupby(filedf, :oldname), &quot;snum&quot;=&gt; (sn-&gt; length(unique(sn)) != 1) =&gt; &quot;ambiguous&quot;);

julia&gt; unique(subset(filedf, &quot;ambiguous&quot;=&gt; identity), [&quot;oldname&quot;, &quot;snum&quot;])
16×6 DataFrame
 Row │ oldname    newname   snum       oldpath                            newpath                            ambiguous
     │ SubStrin…  String    SubStrin…  String                             String                             Bool
─────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 │ FG00004    SEQ01071  S26        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   2 │ FG00004    SEQ01071  S46        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   3 │ FG00005    SEQ01950  S50        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   4 │ FG00005    SEQ01950  S58        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   5 │ FG00016    SEQ01110  S52        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   6 │ FG00016    SEQ01110  S70        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   7 │ FG00017    SEQ02084  S64        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   8 │ FG00017    SEQ02084  S82        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
   9 │ FG00021    SEQ01862  S53        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  10 │ FG00021    SEQ01862  S94        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  11 │ FG02294    SEQ02303  S1         /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  12 │ FG02294    SEQ02303  S34        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  13 │ FG02471    SEQ01727  S16        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  14 │ FG02471    SEQ01727  S89        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  15 │ FG02614    SEQ01971  S10        /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true
  16 │ FG02614    SEQ01971  S9         /grace/sequencing/processed/mgx/…  /grace/sequencing/processed/mgx/…       true</code></pre><p>So for now, we&#39;ll leave these out and deal with them later.</p><h2 id="Renaming"><a class="docs-heading-anchor" href="#Renaming">Renaming</a><a id="Renaming-1"></a><a class="docs-heading-anchor-permalink" href="#Renaming" title="Permalink"></a></h2><pre><code class="language-julia-repl hljs">julia&gt;  for grp in groupby(filedf, &quot;oldname&quot;)
            old = first(grp.oldname)
            @info &quot;Working on $old&quot;
            if any(grp.ambiguous)
                @warn &quot;$old has multiple sequecing results: $(unique(grp.snum)); skipping!&quot;
            else
                @info &quot;Renaming $(old) to $(first(grp.newname))&quot;
                for row in eachrow(grp)
                    @debug &quot;$(row.oldpath) =&gt; $(row.newpath)&quot;
                    mv(row.oldpath, row.newpath)
                end
            end
        end</code></pre><h2 id="Dealing-with-Ambiguities"><a class="docs-heading-anchor" href="#Dealing-with-Ambiguities">Dealing with Ambiguities</a><a id="Dealing-with-Ambiguities-1"></a><a class="docs-heading-anchor-permalink" href="#Dealing-with-Ambiguities" title="Permalink"></a></h2><p>Let&#39;s go back to our ambiguous sequences.</p><pre><code class="language-julia-repl hljs">julia&gt; ambi = subset(filedf, &quot;ambiguous&quot;=&gt; identity);

julia&gt; unique(select(ambi, [&quot;oldname&quot;, &quot;newname&quot;, &quot;snum&quot;]))
16×3 DataFrame
 Row │ oldname    newname   snum
     │ SubStrin…  String    SubStrin…
─────┼────────────────────────────────
   1 │ FG00004    SEQ01071  S26
   2 │ FG00004    SEQ01071  S46
   3 │ FG00005    SEQ01950  S50
   4 │ FG00005    SEQ01950  S58
   5 │ FG00016    SEQ01110  S52
   6 │ FG00016    SEQ01110  S70
   7 │ FG00017    SEQ02084  S64
   8 │ FG00017    SEQ02084  S82
   9 │ FG00021    SEQ01862  S53
  10 │ FG00021    SEQ01862  S94
  11 │ FG02294    SEQ02303  S1
  12 │ FG02294    SEQ02303  S34
  13 │ FG02471    SEQ01727  S16
  14 │ FG02471    SEQ01727  S89
  15 │ FG02614    SEQ01971  S10
  16 │ FG02614    SEQ01971  S9
</code></pre><p>For the first 12 rows (samples <code>FG00004</code>, <code>F00005</code>, <code>FG00016</code>, <code>FG00017</code>, <code>FG0021</code>, and <code>FG02294</code>), the solution is relatively simple. Each of these had 2 aliquots of the same biospecimen sequenced. So in each of those cases, I duplicated the record in the SequencingPrep table (generating a new <code>SEQ</code> id), added the correct S-well number, and put in the old IDs as aliases. Eg, <code>FG00004</code> is now</p><table><tr><th style="text-align: right">uid</th><th style="text-align: right">autonumber</th><th style="text-align: right">biospecimen</th><th style="text-align: right">swell</th><th style="text-align: right">alias</th></tr><tr><td style="text-align: right">SEQ01071</td><td style="text-align: right">1071</td><td style="text-align: right">FG00004</td><td style="text-align: right">S26</td><td style="text-align: right">C0117<em>4F</em>1A</td></tr><tr><td style="text-align: right">SEQ02505</td><td style="text-align: right">2505</td><td style="text-align: right">FG00004</td><td style="text-align: right">S46</td><td style="text-align: right">C0117<em>4F</em>1B</td></tr></table><p>For FG02614, there is only a single file (<code>FG02614_S9_pfams.tsv</code>) that has the S9 identifier. Looking at the contents of that file:</p><pre><code class="nohighlight hljs">❯ head mgx/humann/regroup/FG02614_S9_pfams.tsv
# Gene Family   FG02610_S9_Abundance-RPKs
UNMAPPED        14128546.0
UNGROUPED       29634105.328100212
UNGROUPED|g__Adlercreutzia.s__Adlercreutzia_equolifaciens       17608.29535663359
UNGROUPED|g__Agathobaculum.s__Agathobaculum_butyriciproducens   44303.814025389926
UNGROUPED|g__Akkermansia.s__Akkermansia_muciniphila     118128.883395747
UNGROUPED|g__Alistipes.s__Alistipes_finegoldii  125084.06209765507
UNGROUPED|g__Alistipes.s__Alistipes_indistinctus        24720.017810163998
UNGROUPED|g__Alistipes.s__Alistipes_putredinis  1052709.4440796026
UNGROUPED|g__Alistipes.s__Alistipes_shahii      86567.54527132263</code></pre><p>Based on the header, it looks like it belongs to <code>FG02610</code>. We have all of the files for <code>FG02610</code>, so I&#39;m just going to delete these files (the <code>FG02614_S9</code> ones).</p><p><code>FG02471</code> is weirder. We have all of the relevant files for both <code>S16</code> and <code>S89</code>. Looking at the contents, </p><pre><code class="nohighlight hljs">kevin in vkclab-ada in /lovelace/sequencing/processed on ☁️  (us-east-1)
❯ head -5 mgx/metaphlan/FG02471_S*_profile.tsv
==&gt; mgx/metaphlan/FG02471_S16_profile.tsv &lt;==
#mpa_v30_CHOCOPhlAn_201901
#/home/vklepacc/miniconda3/envs/biobakery3/bin/metaphlan output/kneaddata/FG02471_S16_kneaddata.fastq output/metaphlan/FG02471_S16_profile.tsv --bowtie2out output/metaphlan/FG02471_S16_bowtie2.tsv --samout output/metaphlan/FG02471_S16.sam --input_type fastq --nproc 8 --bowtie2db /pool001/vklepacc/databases/metaphlan
#SampleID       Metaphlan_Analysis
#clade_name     NCBI_tax_id     relative_abundance      additional_species
k__Bacteria     2       99.08787

==&gt; mgx/metaphlan/FG02471_S89_profile.tsv &lt;==
#mpa_v30_CHOCOPhlAn_201901
#/home/vklepacc/miniconda3/envs/biobakery3/bin/metaphlan output/kneaddata/C0477-5F-1A_S89_merged.fastq output/metaphlan/C0477-5F-1A_S89_profile.tsv --bowtie2out output/metaphlan/C0477-5F-1A_S89_bowtie2.tsv --samout output/metaphlan/C0477-5F-1A_S89.sam --input_type fastq --nproc 8
#SampleID       Metaphlan_Analysis
#clade_name     NCBI_tax_id     relative_abundance      additional_species
k__Bacteria     2       99.12309

kevin in vkclab-ada in /lovelace/sequencing/processed on ☁️  (us-east-1)
❯ head -5 mgx/humann/main/FG02471_*genefamilies.tsv
==&gt; mgx/humann/main/FG02471_S16_genefamilies.tsv &lt;==
# Gene Family   FG02471_S16_Abundance-RPKs
UNMAPPED        4356120.0000000000
UniRef90_A0A3E2XRX6     15802.1208367696
UniRef90_A0A3E2XRX6|g__Dorea.s__Dorea_longicatena       15802.1208367696
UniRef90_A5ZYV3 12866.1165754399

==&gt; mgx/humann/main/FG02471_S89_genefamilies.tsv &lt;==
# Gene Family   C0477-5F-1A_S89_Abundance-RPKs
UNMAPPED        3152316.0000000000
UniRef90_A7V4G2 8649.8613270401
UniRef90_A7V4G2|g__Bacteroides.s__Bacteroides_uniformis 8552.3697112122
UniRef90_A7V4G2|unclassified    97.4916158279</code></pre><p>The <code>S89</code> has the old ID in the header (<code>C0477-5F-1A</code>), which is the correct one according to the old database, but it looks like this same ID was sequenced in a different batch (old batch 11) under the ID <code>FG00757</code>. Given the surrounding S-well IDs, it looks like <code>S16</code>, sequenced in (old) batch 17 is correctly attributed to <code>FG02471</code> (which will become <code>SEQ01727</code>), while <code>S89</code> should go with <code>FG00757</code> (which becomes <code>SEQ02512</code>). </p><h2 id="Conclusion"><a class="docs-heading-anchor" href="#Conclusion">Conclusion</a><a id="Conclusion-1"></a><a class="docs-heading-anchor-permalink" href="#Conclusion" title="Permalink"></a></h2><p>This example shows how to download and access the local airtable base, use it to generate a rename mapping, and use that to rename files, taking into account ambiguities.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 28 June 2023 14:48">Wednesday 28 June 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
